### PR TITLE
Make NPCs escape onto tiles with weaker fields

### PR DESCRIPTION
**Pull request by Coolthulhu**
Before:
NPC standing in a 3x3 block of acid would stay there

Now:
NPC will find the tile with lowest sum of dangerous field densities in 3x3 block and move there. Possibly just stumbling around if all tiles are equally bad

Makes it less likely that a NPC will stand in spitter acid.

### CODE REVIEW

File: src/npcmove.cpp [hunk 1]
```diff
@@ -86,14 +86,73 @@ bool clear_shot_reach( const tripoint &from, const tripoint &to )
...
+{
+    std::vector<tripoint> candidates;
+
+    const auto rate_pt = [&who]( const tripoint &p ) {
+        if( !g->m.passable( p ) ) {
+            return INT_MAX;
+        }
+
+        int rating = 0;
+        for( const auto &e : g->m.field_at( p ) ) {
```
Comment (mugling) [inline src/npcmove.cpp]: `std::accumulate`

File: src/npcmove.cpp [hunk 2]
```diff
@@ -86,14 +86,73 @@ bool clear_shot_reach( const tripoint &from, const tripoint &to )
...
+            if( who.is_dangerous_field( e.second ) ) {
+                // @todo Rate fire higher than smoke
+                rating += e.second.getFieldDensity();
+            }
+        }
+
+        return rating;
+    };
+
+    int best_rating = rate_pt( who.pos() );
```
Comment (mugling) [inline src/npcmove.cpp]: Could be refactored as generic STL algorithms:

1. `std::sort` using `rate_pt` as comparator
2. `std::min_element` to find best option
3. `std::remove_if` to eliminate all tiles except the best choices and current tile
4. `random_entry`

Easier to follow and less space for bugs than branching loops

File: src/npcmove.cpp [hunk 3]
```diff
@@ -86,14 +86,73 @@ bool clear_shot_reach( const tripoint &from, const tripoint &to )
...
+            continue;
+        }
+
+        const auto *entry_here = fields_here.findField( e.first );
+        if( entry_here == nullptr || entry_here->getFieldDensity() < e.second.getFieldDensity() ) {
+            return false;
+        }
+    }
+
+    return true;
```
Comment (codemime) [inline src/npcmove.cpp]: `could_move_onto()` (21 lines) can become a one-line function again if you extract `rate_pt()` from `good_escape_direction()` and make it available here (I'd also suggest changing its name due to the wider scope):

```c++
return g->m.passable( p ) && rate_pt( p ) <= rate_pt( pos() );
```

Furthermore, your `rate_pt()` is better than the code above: it takes a sum of fields into account and doesn't bail out after one missing field.

**Comment by mugling** [comment]
These are all would-be-nice changes. I'm trying to unblock as much of the PR queue as possible and given @Coolthulhu maintains the NPC code (and therefore eats their own dog food) I'm going to push this as an acceptable solution to an existing problem

**Pull request merged**